torch/_higher_order_ops/wrap.py

```diff
@@ -110,24 +104,41 @@ def divide_kwargs(kwargs):
         return checkpoint_kwargs, gmod_kwargs
 
     def tag_nodes(self, gmod):
-        # TODO - This needs major investigation. Currently, we are tagging all
-        # the forward nodes as recomputable. However, torch.utils.checkpoint
-        # provides a custom function to selectively recompute. We will have to
-        # figure out how to tag seletively.
         unique_graph_id = next(uid)
         for node in gmod.graph.nodes:
             if node.op in ("call_function", "call_method", "call_module"):
                 node.meta["recompute"] = unique_graph_id
         return gmod
 
     def __call__(self, gmod, *args, **kwargs):
-        if "context_fn" in kwargs:
-            raise RuntimeError("Tagged Activation checkpointing does not support selective checkpointing yet.")
         import torch.fx.traceback as fx_traceback
         from torch.fx import Interpreter
-        gmod = self.tag_nodes(gmod)
-        # Using interpreter allows preservation of metadata through torch.compile stack.
-        with fx_traceback.preserve_node_meta():
-            return Interpreter(gmod).run(*args)
-
-tag_activation_checkpoint = TagActivationCheckpoint()
+        # TODO - This is a temporary sitaution where we have two versions of
+        # checkpointing implemetation. We will converge on one and remove the other.
+        if self.context_fn is not None or torch._functorch.config.functionalize_rng_ops:
+            # Mode 1: WrapActivationCheckpoint
+            log.warning("""
+Detected selective checkpointing is used under torch.compile.
+Please make sure the checkpointed region does not contain:
+
+1. random ops (e.g. torch.dropout)
+2. in-place ops (e.g. torch.relu_)
+""")
+            kwargs["use_reentrant"] = False
+            kwargs["preserve_rng_state"] = False
+            if self.context_fn is not None:
+                kwargs["context_fn"] = self.context_fn
+                # We first tag all nodes as "recompute" in this graph, and then we undo the "recompute" tag
+                # for specific nodes in _CachedTorchDispatchMode.
+                gmod = self.tag_nodes(gmod)
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return checkpoint(Interpreter(gmod).run, *args, **kwargs)
+        else:
+            # Mode 2: TagActivationCheckpoint
+            gmod = self.tag_nodes(gmod)
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return Interpreter(gmod).run(*args)
```

Comment:
we should put SAC+compile logic into TagActivationCheckpoint (and leave the old WrapActivationCheckpoint mode as-is).

Also try to see if full-checkpoint can use the same mode-based `return checkpoint(Interpreter(gmod).run, *args, **kwargs)` logic (by having a special pair of modes for full case).